### PR TITLE
FEXCore: Warn if MDWE is set

### PR DESCRIPTION
**Pull request by Sonicadvance1**
This will result in FEX not being able to allocate executable memory. We can use shared memory in the future to work around this but for now we don't support that as a fix.

### CODE REVIEW

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 1]
```diff
@@ -349,6 +353,19 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
...
 auto CPUBackend::AllocateNewCodeBuffer(size_t Size) -> CodeBuffer {
+#ifndef _WIN32
+// This is a very new Linux 6.3 feature.
+#ifndef PR_GET_MDWE
+#define PR_GET_MDWE 66
+#endif
+  int MDWE = ::prctl(PR_GET_MDWE, 0, 0, 0, 0);
+  if (MDWE != -1 && MDWE != 0) {
+    // If trying to get MDWE didn't result in an error and it had a value set then print an error.
+    // This will mean FEX can't allocate any executable memory.
```
[the remark this answers is not in the document]
Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/CPUBackend.cpp:379]: Spammed out some broken english with an updated comment to contemplate what MDWE is.

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 2]
```diff
@@ -349,6 +353,34 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
 }
 
 auto CPUBackend::AllocateNewCodeBuffer(size_t Size) -> CodeBuffer {
+#ifndef _WIN32
+// This is a very new Linux 6.3 feature.
+// MDWE is short for Memory-Deny-Write-Execute.
+// This is a Linux kernel implementtation of the systemd feature called `MemoryDenyWriteExecute` as well.
+// While the systemd feature is implemented as a seccomp bpf filter, this is implemented entirely in the kernel.
```
Comment (neobrain) [inline FEXCore/Source/Interface/Core/CPUBackend.cpp:360]:
```suggestion
// MDWE (Memory-Deny-Write-Execute) is a new Linux 6.3 feature.
// It's equivalent to systemd'c `MemoryDenyWriteExecute` but implemented entirely in the kernel.
```

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 3]
```diff
@@ -349,6 +353,34 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
...
+//
+// The aim of this feature is to make applications incapable of creating RWX memory mappings. This means as soon as FEX would try
+// to create its RWX JIT memory mappings, mmap would instead return EACCESS. Thus preventing FEX from doing anything JIT related.
```
Comment (neobrain) [inline FEXCore/Source/Interface/Core/CPUBackend.cpp:363]:
```suggestion
// MDWE prevents applications from creating RWX memory mappings. This prevents FEX from doing anything JIT related, as FEX uses RWX for JIT memory mappings.
```

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 4]
```diff
@@ -349,6 +353,34 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
...
+//
+// A workaround is to call mprotect every time we need to write or modify code or to use a memory mirror where one half is mapped as RW and the other
+// is RX.
```
Comment (neobrain) [inline FEXCore/Source/Interface/Core/CPUBackend.cpp:366]:
```suggestion
// A potential workaround to make FEX work with MDWE is to call mprotect every time we need to write or modify code. Alternatively, FEX could use a memory mirror where one half is mapped as RW and the other is RX.
```

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 5]
```diff
@@ -349,6 +353,34 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
...
+//
+// Once MDWE is enabled with the prctl, the feature is sealed and it can /NOT/ be turned off.
+//
+// To check if MDWE is enabled, use prctl to with the option of `PR_GET_MDWE`.
+// First ensure that -1 isn't returned, to make sure the kernel supports it.
+// Then make sure the value returned is zero. If the value is /NOT/ zero that means MDWE is enabled and RWX can't be mapped.
```
Comment (neobrain) [inline FEXCore/Source/Interface/Core/CPUBackend.cpp:372]:
```suggestion
// Status of MDWE is queried through prctl using `PR_GET_MDWE`:
// -1: The kernel doesn't support MDWE
// 0: MDWE is supported but disabled
// >0: MDWE is enabled, hence prohibiting RWX mappings
```

File: FEXCore/Source/Interface/Core/CPUBackend.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (neobrain) [reply FEXCore/Source/Interface/Core/CPUBackend.cpp:379]: Thanks! The two-line comment here is redundant now.

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 6]
```diff
@@ -349,6 +353,29 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
 }
 
 auto CPUBackend::AllocateNewCodeBuffer(size_t Size) -> CodeBuffer {
+#ifndef _WIN32
+// MDWE (Memory-Deny-Write-Execute) is a new Linux 6.3 feature.
+// It's equivalent to systemd'c `MemoryDenyWriteExecute` but implemented entirely in the kernel.
```
Comment (neobrain) [inline FEXCore/Source/Interface/Core/CPUBackend.cpp:358]:
```suggestion
// It's equivalent to systemd's `MemoryDenyWriteExecute` but implemented entirely in the kernel.
```

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 7]
```diff
@@ -349,6 +353,29 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
...
+//
+// MDWE prevents applications from creating RWX memory mappings. This prevents FEX from doing anything JIT related, as FEX uses RWX for JIT memory mappings.
+//
+// A potential workaround to make FEX work with MDWE is to call mprotect every time we need to write or modify code. Alternatively, FEX could use a memory mirror where one half is mapped as RW and the other is RX.
```
Comment (neobrain) [inline FEXCore/Source/Interface/Core/CPUBackend.cpp:362]:
```suggestion
// A potential workaround to make FEX work with MDWE is to call mprotect every time we need to write or modify code.
// Alternatively, FEX could use a memory mirror where one half is mapped as RW and the other is RX.
```

File: FEXCore/Source/Interface/Core/CPUBackend.cpp [hunk 8]
```diff
@@ -349,6 +353,29 @@ auto CPUBackend::GetEmptyCodeBuffer() -> CodeBuffer * {
 }
 
 auto CPUBackend::AllocateNewCodeBuffer(size_t Size) -> CodeBuffer {
+#ifndef _WIN32
+// MDWE (Memory-Deny-Write-Execute) is a new Linux 6.3 feature.
+// It's equivalent to systemd'c `MemoryDenyWriteExecute` but implemented entirely in the kernel.
+//
+// MDWE prevents applications from creating RWX memory mappings. This prevents FEX from doing anything JIT related, as FEX uses RWX for JIT memory mappings.
```
Comment (neobrain) [inline FEXCore/Source/Interface/Core/CPUBackend.cpp:360]:
```suggestion
// MDWE prevents applications from creating RWX memory mappings.
// This prevents FEX from doing anything JIT related, as FEX uses RWX for JIT memory mappings.
```

**Review by neobrain — APPROVED** [review]

**Pull request merged**